### PR TITLE
feat: add permission to manage all projects gf-248

### PR DESCRIPTION
**Pull request by Fjortis**
Added manage_all_projects permission (name: "Manage All Projects") to the permissions table.
Users with this permission should be able to see projects item in the sidebar and view, create, edit (including project groups), and delete projects.

Users with this permission can see the "Projects" item in the sidebar.
The options button is visible for such users on the Specific Project page.
User have an access to view the "Access Managment" page though the Options button on the Specific Project page.
User can edit, delete projects.
User doe not see the "Access Managment" page on the sidebar.

No permission
Mana all projects no permission

With Manage all projects permission
Manage all perm, project view

Manage All Projects permission, menu
Project menu open

Manage All Projects permission update menu
Update open

### CODE REVIEW

**Comment by perekotypole** [comment]
Name PR as `feat: add permission to manage all projects gf-248` to avoid PR lint error

**Comment by Fjortis** [comment]
I have quite a few adjustments i need to do, i see that i need to refactor the project page, adjust modal behavior onclode for edit, and i need to adjust permissions to handle multiple permissions in routing, currently i am not sure about the best approach for that. i am looking into it

File: apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx
Comment (s1rserg) [inline apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx:1]:
```suggestion
import { useCallback } from "~/libs/hooks/hooks.js";
```

File: apps/frontend/src/index.tsx
Comment (perekotypole) [inline apps/frontend/src/index.tsx:49]: Do we need this permission for access management?

File: apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx]: Had now been adjusted, thanks!

File: apps/frontend/src/index.tsx
Comment (Fjortis) [reply apps/frontend/src/index.tsx:49]: From what i understood the person with Manage all Projects should be able to access it, but not the sidebar for access management, i believe in the future the managing of Project groups will be handled there, we do not have anything of it in project specification, so i left it like that, until further changes is implemented, perhaps it is better to remove it until start working on project groups?

File: apps/frontend/src/pages/project/project.tsx
Comment (kolibri753) [inline apps/frontend/src/pages/project/project.tsx:82]: Why was hasProject removed in favor of checking project directly? I think using hasProject improves readability and matches the structure of the app.

File: apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx
Comment (kolibri753) [inline apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx:14]: Why is it optional?

Comment (AnnKasian) [reply apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx:14]: It's not merged yet)
It's the project access management page, not the main access

File: apps/frontend/src/pages/project/project.tsx
Comment (AnnKasian) [reply apps/frontend/src/pages/project/project.tsx:82]: Agree, @Fjortis have a look at this [comment](https://github.com/BinaryStudioAcademy/bsa-2024-gitfit/pull/171#discussion_r1742413987)

Comment (Fjortis) [reply apps/frontend/src/pages/project/project.tsx]: I adjusted this, thanks for the reference

File: apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx:14]: I still adjusted it, it doesn't really matter since it will still be changed after the project access management page implementation.

File: apps/frontend/src/index.tsx
Comment (liza-veis) [reply apps/frontend/src/index.tsx:49]: user with MANAGE_ALL_PROJECTS permission, won't be able to access root AccessManagement page, so we need to remove this

Comment (Fjortis) [reply apps/frontend/src/index.tsx:49]: This has now been adjusted

File: apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx
Comment (perekotypole) [inline apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx:51]: Does it work as we need with `NavLink` component and `onClick` in inside component? Won't the actions be triggered at the same time and cause a conflict?

File: apps/frontend/src/pages/project/project.tsx
Comment (perekotypole) [inline apps/frontend/src/pages/project/project.tsx:30]: Perhaps it would be better to either remove it and make `onClick` optional, or to navigate using the function?

File: apps/frontend/src/pages/project/project.tsx
Comment (perekotypole) [inline apps/frontend/src/pages/project/project.tsx:56]: Why not put it in a separate `useEffect`?

File: apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx:51]: I split them into two useEffects to keep the logic cleaner and easier to follow. One handles fetching the project, and the other manages the modal state.

File: apps/frontend/src/pages/project/project.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/project.tsx:30]: I’ve made onClick optional, so now we can use it when needed or rely on NavLink for navigation. It keeps things flexible.

File: apps/frontend/src/pages/project/project.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/project.tsx:56]: I split them into two useEffects to keep the logic cleaner and easier to follow. One handles fetching the project, and the other manages the modal state.

File: apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/libs/components/project-menu/project-menu.tsx:51]: I made sure we’re not using onClick with NavLink to avoid conflicts. Now we handle navigation and actions separately.

**Comment by liza-veis** [comment]
Looks good to me, we need to wait before #257 merged

File: apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx
Comment (liza-veis) [inline apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx:34]: we will have multiple permissions here
```suggestion
	const hasManageAccessPermission = checkHasPermission(
```

File: apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx
Comment (liza-veis) [inline apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx:41]: we need to render only Manage Access item conditionally, not the whole menu
```suggestion
```

File: apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx
Comment (liza-veis) [inline apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx:48]: We need to do something similar to the approach in the Button component. We need to pass onClick or href to menu item, and then render NavLink if needed inside it
```suggestion
				<MenuItem iconName="access" label="Manage Access" href={projectAccessManagementRoute} />
```

**Review by liza-veis — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx]: This has now been implemented, i've made `href `and onclick optional prop for `menuItem`

File: apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx]: Now it only rendered the `menuItem ` Manage Access depending on `hasManageAccessPermission`

File: apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx
Comment (Fjortis) [reply apps/frontend/src/pages/project/libs/components/project-details-menu/project-details-menu.tsx]: Changed the naming

**Review by what1s1ove — APPROVED** [review]
lgtm after Lisa's comments

**Comment by Fjortis** [comment]
Hmmm, seems like something is off, on my final testing after last merge, it wont accept anything with multiple permissions, both frontend and backend, i need to take a look into this...

Any idea what went wrong?

**Review by liza-veis — APPROVED** [review]

**Comment by liza-veis** [comment]
@Fjortis waiting for update from you regarding the issue you mentioned

**Comment by Fjortis** [comment]
> @Fjortis waiting for update from you regarding the issue you mentioned
I'm struggling with a permissions issue where for example a sidebar item requires multiple permissions, but it's only shown if the user has all the permissions. I need to adjust it so the item is shown if the user has at least one of the required permissions, but I'm unsure how to best handle that.

image

**Comment by liza-veis** [comment]
The update in the `checkHasPermission ` helper looks correct to me. Regarding permissions delay, the logic should be that we load groups with authenticated user, and until we have him we should show loader and don't check permissions. At least this is how it should work.

**Review by liza-veis — APPROVED** [review]

**Pull request merged**
